source/nsdynmemLIB/nsdynmemLIB.c

```diff
@@ -20,102 +20,132 @@
 #include <stdlib.h>
 #include "ns_list.h"
 
-void (*heap_failure_callback)(heap_fail_t);
-
 #ifndef STANDARD_MALLOC
-static int *heap_main = 0;
-static int *heap_main_end = 0;
-static uint16_t heap_size = 0;
-
 typedef enum mem_stat_update_t {
     DEV_HEAP_ALLOC_OK,
     DEV_HEAP_ALLOC_FAIL,
     DEV_HEAP_FREE,
 } mem_stat_update_t;
 
-
-static mem_stat_t *mem_stat_info_ptr = 0;
-
 typedef struct {
     ns_list_link_t link;
 } hole_t;
 
-static NS_LIST_DEFINE(holes_list, hole_t, link);
+typedef int ns_mem_word_size_t; // internal signed heap block size type
+
+/* struct for book keeping variables */
+typedef struct ns_mem_book {
+    ns_mem_word_size_t     *heap_main;
+    ns_mem_word_size_t     *heap_main_end;
+    mem_stat_t *mem_stat_info_ptr;
+    void (*heap_failure_callback)(heap_fail_t);
+    NS_LIST_HEAD(hole_t, link) holes_list;
+    ns_mem_heap_size_t heap_size;
+} ns_mem_book_t;
+
+static ns_mem_book_t *static_context; // heap pointer for original "ns_" API use
 
 // size of a hole_t in our word units
-#define HOLE_T_SIZE ((sizeof(hole_t) + sizeof(int) - 1) / sizeof(int))
+#define HOLE_T_SIZE ((sizeof(hole_t) + sizeof(ns_mem_word_size_t) - 1) / sizeof(ns_mem_word_size_t))
 
-static NS_INLINE hole_t *hole_from_block_start(int *start)
+static NS_INLINE hole_t *hole_from_block_start(ns_mem_word_size_t *start)
 {
     return (hole_t *)(start + 1);
 }
 
-static NS_INLINE int *block_start_from_hole(hole_t *start)
+static NS_INLINE ns_mem_word_size_t *block_start_from_hole(hole_t *start)
 {
-    return ((int *)start) - 1;
+    return ((ns_mem_word_size_t *)start) - 1;
 }
 
-
-static void heap_failure(heap_fail_t reason)
+static void heap_failure(void (*callback)(heap_fail_t), heap_fail_t reason)
```

Comment:
Neater and more efficient to pass the book pointer here, rather than the function pointer.